api/classes/base_controller.js

```diff
@@ -11,133 +13,165 @@ function defaultFormatResponse(model) {
-// we do it in multiple turns of the event loop, so that other requests
-// don't get blocked.
-BaseController.prototype.getAllAsTar = function(req, reply) {
-  var files = req.pre.records.models;
-  var tarStream = Tar.pack();
-  var model = this.Model;
-
-  function processFile(file) {
-    return model.query({
-      where: {
-        id: file.get('id')
-      },
-      columns: ['buffer']
-    }).fetch().then(function(model) {
-      return new Promise(function(resolve) {
-        setImmediate(function() {
-          tarStream.entry({ name: file.get('path') }, model.get('buffer'));
-          resolve();
+    .catch(Errors.generateErrorResponse);
+
+    reply(result);
+  }
+
+  getOne(request, reply) {
+    const record = request.pre.records.models[0].toJSON();
+
+    reply(
+      request.generateResponse(
+        this.formatResponseData(record)
+      ).code(200)
+    );
+  }
+
+  getAll(request, reply) {
+    const records = request.pre.records.toJSON();
+
+    reply(
+      request.generateResponse(
+        records.map(this.formatResponseData)
+      )
+    );
+  }
+
+  getAllAsMeta(request, reply) {
+    reply(
+      request.generateResponse(
+        request.pre.records.toJSON()
+      )
+    );
+  }
+
+  // NOTE: creating the tarball for a project can be a lengthy process, so
+  // we do it in multiple turns of the event loop, so that other requests
+  // don't get blocked.
+  getAllAsTar(request, reply) {
+    const files = request.pre.records.models;
+    const tarStream = Tar.pack();
+    const Model = this.Model;
+
+    function processFile(file) {
+      return Model.query({
+        where: {
+          id: file.get(`id`)
+        },
+        columns: [`buffer`]
+      })
+      .fetch()
+      .then(function(model) {
+        return new Promise(function(resolve) {
+          setImmediate(function() {
+            tarStream.entry(
+              {
+                name: file.get(`path`)
+              },
+              model.get(`buffer`)
+            );
+
+            resolve();
+          });
         });
       });
+    }
+
+    setImmediate(function() {
+      Promise.map(files, processFile, { concurrency: 2 })
+      .then(function() { return tarStream.finalize(); })
+      .catch(Errors.generateErrorResponse);
     });
+
+    // Normally this type would be application/x-tar, but IE refuses to
+    // decompress a gzipped stream when this is the type.
+    reply(tarStream)
+    .header(`Content-Type`, `application/octet-stream`);
+  }
+
+  // `formatResponse` is an optional processing function that can be passed
+  // in to modify what is sent in the response body. If no function is
+  // provided, the full model for the current method is used in the response.
+  update(request, reply, formatResponse) {
+    const requestData = this.formatRequestData(request);
+
+    formatResponse = typeof formatResponse === `function` ? formatResponse : defaultFormatResponse;
```

Comment:
same note - this might be better as an explicit "don't do anything unless typeof is not function"